### PR TITLE
Fix potential NRE in GenericSpecializationPartCreationInfo

### PR DESCRIPTION
**Pull request by Alex4414**
Method GenericSpecializationPartCreationInfo.CanSpecialize checks that arrays genericParameterConstraints and genericParameterAttributes are both null or have expected lengths only. But it doesn't check that one of genericParameterConstraints and genericParameterAttributes is null, but other is not null and have expected length. In this case it will be NullReferenceException.

Now these arrays are either saved to metadata together or not saved together: 

```
internal static IDictionary<string, object?> GetPartMetadataForType(this Type type, CreationPolicy creationPolicy)
{
...
                if (hasConstraints)
                {
                    dictionary.Add(CompositionConstants.GenericParameterConstraintsMetadataName, genericParameterConstraints);
                    dictionary.Add(CompositionConstants.GenericParameterAttributesMetadataName, genericParameterAttributes);
                }
...
}
```

So in case when one if it is null and other is not null, we can return false instead of NullReferenceException.

Found by Linux Verification Center (linuxtesting.org) with SVACE.

### CODE REVIEW

**Comment by Alex4414** [comment]
Hi @adamsitnik, can you try to find a reviewer offline, please?

**Review by adamsitnik — COMMENTED** [review]
Thank you for your contribution @Alex4414 ! The change looks good, but since we are making modifications lets remove the `!` that became redundant few lines below.

File: src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ReflectionModel/GenericSpecializationPartCreationInfo.cs [hunk 1]
```diff
@@ -562,12 +562,17 @@ public static bool CanSpecialize(IDictionary<string, object?> partMetadata, Type
                 return true;
             }
 
-            if ((genericParameterConstraints != null) && (genericParameterConstraints.Length != partArity))
+            if ((genericParameterConstraints == null) || (genericParameterAttributes == null))
```
[the remark this answers is not in the document]
Comment (Alex4414) [reply src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ReflectionModel/GenericSpecializationPartCreationInfo.cs:565]: Sure, done!

**Comment by Alex4414** [comment]
@adamsitnik Thank you, sure, I removed it

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you @Alex4414 !

**Pull request merged**